### PR TITLE
[29] Cluster description sentences

### LINKED ISSUE #29: Cluster sentence embeddings
Find clusters of semantically similar sentences, cluster them and explore how they are distributed across subjects.

Proposed steps:
- Split course descriptions into sentences
- Remove very short sentences
- Embed each sentence
- Cluster the embeddings for each subject
- Investigate the content and meaning of each cluster
- Explore the distribution and representation of each cluster (e.g. relative specialisation of clusters for each school)
- Plot charts showing the distribution of cluster representation across schools
- Discuss how these might be linked to outcomes based on gender

### PR DESCRIPTION
**Pull request by georgerichardson**
---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

closes #29

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -17,3 +17,6 @@ python-dotenv
 seaborn
 textstat
 tabulate
+umap-learn
+scikit-learn
+toolz
```
Comment (Jack-0-0) [inline requirements.txt:22]: Need to add hdbscan to the requirements

**Review by Jack-0-0 — COMMENTED** [review]
Thanks @georgerichardson.

The code looks good, there is just a missing requirement.

I tried to run the code. `comp_sci_gender_bias/pipeline/sentence_clusters/create_clusters.py` ran fine and produced the expected outputs.

When running `comp_sci_gender_bias/analysis/sentence_clusters.py` it has a KeyError for  `cluster_type` when running line 70 in function `category_token_distribution`.

Do you think it would be helpful to also add a readme on how you manually inspected the clusters and grouped them?

**Comment by georgerichardson** [comment]
> The code looks good, there is just a missing requirement.

Fixed, thanks.

> I tried to run the code. comp_sci_gender_bias/pipeline/sentence_clusters/create_clusters.py ran fine and produced the expected outputs.

👍 

> When running comp_sci_gender_bias/analysis/sentence_clusters.py it has a KeyError for cluster_type when running line 70 in function category_token_distribution.

I expect this is because you had just run `comp_sci_gender_bias/pipeline/sentence_clusters/create_clusters.py` which created new clusters. If there were fewer clusters than when I grouped them, then this would trigger a KeyError. You could test it by checking the output sentence data csv files back out from the branch. I don't like how fragile this is tbh.

> Do you think it would be helpful to also add a readme on how you manually inspected the clusters and grouped them?

I have added a small line to the readme which points to a notebook with the process that I followed.

**Review by Jack-0-0 — APPROVED** [review]
> > When running comp_sci_gender_bias/analysis/sentence_clusters.py it has a KeyError for cluster_type when running line 70 in function category_token_distribution.
> 
> I expect this is because you had just run `comp_sci_gender_bias/pipeline/sentence_clusters/create_clusters.py` which created new clusters. If there were fewer clusters than when I grouped them, then this would trigger a KeyError. You could test it by checking the output sentence data csv files back out from the branch. I don't like how fragile this is tbh.
> 

Ah, that makes sense. It runs fine with the original files in the outputs directory 🙂

> > Do you think it would be helpful to also add a readme on how you manually inspected the clusters and grouped them?
> 
> I have added a small line to the readme which points to a notebook with the process that I followed.

Great, thanks!

The PR looks good to merge.

**Pull request merged**